src/crimson/osd/osd_operations/client_request.cc

```diff
@@ -59,38 +59,53 @@ seastar::future<> ClientRequest::start()
   IRef opref = this;
   return crimson::common::handle_system_shutdown(
     [this, opref=std::move(opref)]() mutable {
-    return with_blocking_future(handle.enter(cp().await_map))
-    .then([this]() {
-      return with_blocking_future(osd.osdmap_gate.wait_for_map(m->get_min_epoch()));
-    }).then([this](epoch_t epoch) {
-      return with_blocking_future(handle.enter(cp().get_pg));
-    }).then([this] {
-      return with_blocking_future(osd.wait_for_pg(m->get_spg()));
-    }).then([this, opref=std::move(opref)](Ref<PG> pgref) {
-      return seastar::do_with(
-	std::move(pgref), std::move(opref), [this](auto& pgref, auto& opref) {
-	  PG &pg = *pgref;
-	  return with_blocking_future(
-	    handle.enter(pp(pg).await_map)
-	  ).then([this, &pg]() mutable {
+    return seastar::repeat([this, opref]() mutable {
+      return with_blocking_future(handle.enter(cp().await_map))
+      .then([this]() {
+	return with_blocking_future(osd.osdmap_gate.wait_for_map(m->get_min_epoch()));
+      }).then([this](epoch_t epoch) {
+	return with_blocking_future(handle.enter(cp().get_pg));
+      }).then([this] {
+	return with_blocking_future(osd.wait_for_pg(m->get_spg()));
+      }).then([this, opref](Ref<PG> pgref) {
+	    PG &pg = *pgref;
 	    return with_blocking_future(
-	      pg.osdmap_gate.wait_for_map(m->get_map_epoch()));
-	  }).then([this, &pg](auto map) mutable {
-	    return with_blocking_future(
-	      handle.enter(pp(pg).wait_for_active));
-	  }).then([this, &pg]() mutable {
-	    return with_blocking_future(pg.wait_for_active_blocker.wait());
-	  }).then([this, &pgref]() mutable {
-	    if (m->finish_decode()) {
-	      m->clear_payload();
-	    }
-	    if (is_pg_op()) {
-	      return process_pg_op(pgref);
-	    } else {
-	      return process_op(pgref);
-	    }
-	  });
-	});
+	      handle.enter(pp(pg).await_map)
+	    ).then([this, &pg]() mutable {
```

Comment:
[off topic] why can we assume that pg is still available in this continuation. as we are merely capturing a reference of it.

Reply:
I moved pgref into a later lambda, so I think this pg should be available within the lifetime of this lambda. Am I right?

Reply:
yeah, i just noticed 
```c++
pgref=std::move(pgref)
```
thanks for explaining this to me!